Augment cf_paths: adding path use strategy and dynamic path choosing logic

tiden0614 opened the pull request:
### Purpose of change
We are prototyping to use rocks db as our next-gen storage engine. One problem that we have encountered is that rocksdb doesn't provide a good way to utilize multiple disk mounts evenly on our hosts. We didn't really like the raid option (which was mentioned in the rocksdb FAQ): our current prod software is running on these hosts with multiple equal-sized disk mounts; using a raid would require a risky disk setup before the data migration.

This change introduces a path use strategy when creating output files. It is built on top the already-existing feature of cf_paths. 

### Major changes
* Added a set of classes in column_family.h to represent different path picking strategies
* Added an "db_path_use_strategy" option in ColumnFamilyOptions and DBOptions
* Modified the FileMetadata struct: use 2 separate integers to represent file number and path id
* Modified the ColumnFamilyData::CompactRange method signature. Gave a default value to the output_path_id argument and use -1 to represent dynamic path id.
* Modified the flush: instead of always using path 0, now it dynamically chooses a path to create the output file
* Modified the compaction: instead of using the pre-chosen path for all output files in a compaction, dynamically choose a path for every single output file.
* Modified the java interface to include the db_path_use_strategy options
* Modified the java interface to include cf_paths options, which were not added in previous changes
* Tweaked the Makefile in a few places where it failed compiling code on our machines
* Removed the limitation of 4 path_ids in manifest serialization

### Testing
* All unit tests passed
* We have been doing some basic load tests with these changes in our test environment (involving a small fleet consisting of dozens of hosts) for a few weeks and haven't discover major issues.

tiden0614 commented:
Apologies first for the large bundle of changes. We've been buffering quite a few tweaks for a while. I think it will get better for our future contributions.

siying commented:
Thank you for your contribution! While we still haven't got bandwidth to look into this giant PR, it's on our list.

tiden0614 commented:
@siying Thanks! This is an important feature for us so we'd really like it to get merged.

I noticed that I've broken 2 universal compaction tests with the change and I'm working on fixing those (I changed the way universal compaction calculating the size of compaction files in subtle ways).

tiden0614 commented:
@siying Hey I have updated the CR the PR to a relatively good state to be reviewed. The previous broken test cases have been fixed. Sorry for it taking me so long to update the PR.

I would appreciate it if you guys can allocate some bandwidth to take a look at this change.

siying commented on db/compaction/compaction_job.cc:
```diff
@@ -567,6 +566,31 @@ void CompactionJob::GenSubcompactionBoundaries() {
+
+  // iterate all output files and collect their path ids into a set
+  auto sub_compact_itr = compact_->sub_compact_states.begin();
+  for (; sub_compact_itr != compact_->sub_compact_states.end(); sub_compact_itr++) {
```
Nit, can we use do `for (auto& sub_compact :  compact_->sub_compact_states)`?

tiden0614 replied:
Fixed

siying commented on db/compaction/compaction_job.cc:
```diff
@@ -567,6 +566,31 @@ void CompactionJob::GenSubcompactionBoundaries() {
+  // fsync all output file path ids
+  Status ret = Status::OK();
+  auto path_id_itr = path_ids.begin();
+  for (; path_id_itr != path_ids.end(); path_id_itr++) {
```
Same nit here. Can we do `for (uint32_t path_id : path_ids)`?

tiden0614 replied:
Fixed

siying commented on db/compaction/compaction.cc:
```diff
@@ -207,11 +207,12 @@ Compaction::Compaction(VersionStorageInfo* vstorage,
                        CompressionOptions _compression_opts,
                        uint32_t _max_subcompactions,
                        std::vector<FileMetaData*> _grandparents,
+                       DbPathSupplier* db_path_supplier,
```
It looks like that we are passing ownership to Compaction class. If that is the case, the best practice is to use `std::unique_ptr<DbPathSupplier>&&`.

tiden0614 replied:
Changed to std::unique_ptr<DbPathSupplier>&& for both compaction and flush job.

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+  }
+
+ protected:
+  const ImmutableCFOptions ioptions_;
```
Why do we need a copy of the option? This is a large object. Can we use `const ImmutableCFOptions&` instead?

tiden0614 replied:
I changed the supplier objects to only make copy to things they actually need instead of the entire option objects.

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+  Status FsyncDbPath(uint32_t path_id) const;
+
+  virtual uint32_t GetPathId(int level) const {
+    UNUSED(level);
```
The practice followed in this code base is to declare it as `int /* level */`

tiden0614 replied:
Fixed

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+  }
+
+ protected:
+  const ImmutableCFOptions ioptions_;
```
Same here. `uint32_t /* path_id */, int /*output_level*/`.

In fact, do we need to define GetPathId() and AcceptPathId() here at all? Can we instead make them ` = 0`?

tiden0614 replied:
Fixed. Also took your advice and make them `=0`.

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+    : DbPathSupplier(ioptions), path_id_(path_id) {}
+
+  uint32_t GetPathId(int level) const override {
+    UNUSED(level);
```
The convenstion is `int /* level */`.

tiden0614 replied:
Fixed

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+
+ private:
+  uint64_t file_size_;
+  const MutableCFOptions moptions_;
```
Do we need to make a copy?

tiden0614 replied:
Changed to only copy a few integers.

siying commented on db/db_path_supplier.h:
```diff
@@ -0,0 +1,127 @@
+      uint32_t path_id, int output_level) const override;
+
+private:
+  const MutableCFOptions moptions_;
```
Do we need to make a copy?

tiden0614 replied:
Changed to only copy a few integers.

siying commented on db/flush_job.h:
```diff
@@ -81,6 +81,7 @@ class FlushJob {
   void Cancel();
   TableProperties GetTableProperties() const { return table_properties_; }
   const autovector<MemTable*>& GetMemTables() const { return mems_; }
+  const FileMetaData GetOutputFileMd() { return meta_; };
```
Do we need to make a copy of FileMetaData? can it return `const FileMetaData&`. In fact, I don't even know whether FileMetaData is copyable.
If we only need to figure out output path, maybe a function to return output path is easier to get right.

tiden0614 replied:
Changed to only return output path id.

siying commented on db/version_edit.h:
```diff
@@ -34,20 +30,22 @@ extern uint64_t PackFileNumberAndPathId(uint64_t number, uint64_t path_id);
   TableReader* table_reader;
-  uint64_t packed_number_and_path_id;
+  uint32_t path_id;
+  uint64_t file_number;
```
Do we have to waste this 8 bytes? Can you explain why it is absolutely needed?

tiden0614 replied:
The motivation for isolating the path_id to its own integer is that: the current way of packing path id and file number only gives a limit of `4` for path id.

I didn't realize that `FileDescriptor` can leave a big memory footprint, so took the simple approach of giving path id its own integer to avoid the casting.

Glad you asked. I made a change to give 2 more bits to path id so that the upper limit now goes up to `16`, which should be sufficient for majority cases. However, it does shrink the upper limit of file numbers quite a bit. It is still a very large upper limit so I'm assuming that it should be fine.

Let me know if this works.

siying commented on include/rocksdb/options.h:
```diff
@@ -78,6 +78,50 @@ enum CompressionType : unsigned char {
+  // Note that the target_size attr in DbPath will not be useful if this
+  // strategy is chosen.
+  kRandomlyChoosePath = 0x01,
+};
```
Do you consider to make DbPathSupplier a public interface, and options take a factory class for it? If the factory is empty, then we fall back to the build-in behavior, otherwise, users can implement it in a way they want?

It can be done just like FlushBlockPolicy and FlushBlockBySizePolicyFactory. We have similar data structures like that.

tiden0614 replied:
I really like this idea and I can see ourselves benefit from this more flexible approach. The ability to expand functionality without sending PR or forking Rocksdb is great.

The following reasons, however, are why I'm not already doing it:
1. I don't know how should we deal with the Java interface, which we plan to use in our application
2. I'm not entirely sure how should the factory interface look like. The current use cases will at least require the factory to have a method 
```cpp
virtual std::unique_ptr<DbPathSupplier> produce(
    const ImmutableOptions& ioptions, 
    const MutableOptions& moptions, 
    uint64_t estimated_file_size)
```
The `estimated_file_size` comes from the universal compaction path choosing logic. The output size is estimated upfront before the compaction actually happens (where `GetPathId()` is called on the supplier). This does feel a bit awkward because `estimated_file_size` is only useful for universal compaction but is added everywhere. There maybe another use case that requires another argument on the factory, thus breaking client side code.

Do you have suggestions for the above confusions?

tiden0614 commented:
@siying this pr has been updated to a ready state to be reviewed. Thanks for allocating bandwidth on this change.

I addressed comments from last revision.

tiden0614 commented:
ping

tiden0614 commented:
@siying appreciate a lot if the team can allocate some cycles review this change. We would like this feature to be merged into rocksdb so that we can continue to use it in a larger scale. 

It would be nice to provide an update, even if it's just an estimation of when we can get some traction makes us happy.

tiden0614 commented:
ping

tiden0614 commented:
@siying ping

tiden0614 commented:
ping

tiden0614 commented:
ping
